react-components/src/query/useSearchMappedEquipmentAssetMappings.tsx

```diff
@@ -27,53 +28,92 @@ export type ModelMappingsWithAssets = ModelMappings & {
   query: string,
   models: AddModelOptions[],
   limit: number = 100,
   userSdk?: CogniteClient
-): UseQueryResult<Asset[]> => {
+): UseInfiniteQueryResult<InfiniteData<AssetPage[]>, Error> => {
   const sdk = useSDK(userSdk);
-  const modelsKey = models.map((model) => [model.modelId, model.revisionId]);
-  const { data: assetMappings, isFetched } = useAllMappedEquipmentAssetMappings(models, sdk);
+  const { data: assetMappingList, isFetched } = useAssetMappedNodesForRevisions(
+    models.map((model) => ({ ...model, type: 'cad' }))
+  );
+  const allAssetMappings = useAllMappedEquipmentAssetMappings(models, sdk);
 
-  return useQuery({
-    queryKey: ['reveal', 'react-components', 'search-mapped-asset-mappings', query, modelsKey],
-    queryFn: async () => {
+  return useInfiniteQuery({
+    queryKey: [
+      'reveal',
+      'react-components',
+      'search-mapped-asset-mappings',
+      query,
+      ...models.map((model) => [model.modelId, model.revisionId])
+    ],
+    queryFn: async ({ pageParam }: { pageParam: string | undefined }) => {
+      if (allAssetMappings.data === undefined) {
+        return { assets: [], nextCursor: undefined };
+      }
       if (query === '') {
-        const mappedAssets =
-          assetMappings?.pages
+        const assets = allAssetMappings.data?.pages.flatMap((modelWithAssets) =>
+          modelWithAssets
+            .map((modelWithAsset) =>
+              modelWithAsset.modelsAssets.flatMap((modelsAsset) => modelsAsset.assets)
+            )
             .flat()
-            .map((item) => item.assets)
-            .flat() ?? [];
-        return mappedAssets;
+        );
+        return { assets, nextCursor: undefined };
       }
-
-      const searchedAssets = await sdk.assets.search({ search: { query }, limit: 1000 });
-      const assetMappingsWithSearch = await getAssetMappingsByModels(
-        sdk,
-        models,
+      if (assetMappingList === undefined) {
+        return { assets: [], nextCursor: undefined };
+      }
+      const assetsResponse = await getAssetsList(sdk, {
+        query,
         limit,
-        searchedAssets.map((asset) => asset.id)
-      );
+        cursor: pageParam
+      });
+
+      const assets = assetsResponse.items.filter((asset) => asset !== undefined);
 
-      const assetMappingsSet = createAssetMappingsSet(assetMappingsWithSearch);
-      const filteredSearchedAssets = searchedAssets.filter((asset) =>
-        assetMappingsSet.has(asset.id)
-      );
+      const assetMappingsWithSearch = assetMappingList.flatMap((mapping) => {
+        return mapping.assetMappings.filter((assetMapping) =>
+          assets.map((asset) => asset.id).includes(assetMapping.assetId)
+        );
+      });
 
-      return filteredSearchedAssets;
+      const assetMappingsSet = new Set(assetMappingsWithSearch.map((mapping) => mapping.assetId));
+      const filteredSearchedAssets = assets.filter((asset) => assetMappingsSet.has(asset.id));
```

Comment:
I don't understand what's happening here :see_no_evil: We first get all asset mappings for the revisions (at the top of this hook), and then we find the overlap between the fetched asset mappings and the fetched assets (with an O(n^2) algorithm), and then we construct a set from that overlap and finally filter the fetched assets based on that set. It seems that this set could be created from the `assetMappingList` up-front, and the assets can be filtered against it on one line here?